themes/bootstrap3/js/common.js

```diff
@@ -154,46 +161,58 @@ var VuFind = (function VuFind() {
       }
     }
   };
-  var icon = function icon(name, attrs = {}) {
+
+  /**
+   * Get an icon identified by a name.
+   *
+   * @param {string} name          Name of the icon to create
+   * @param {Object} attrs         Object containing attributes,
+   *                               key is the attribute of an HTMLElement,
+   *                               value is the values to add for the attribute.
+   * @param {bool}   returnElement [Optional] Should the function return an HTMLElement.
+   *                               Default is false.
+   *
+   * @returns {string|HTMLElement}
+   */
+  var icon = function icon(name, attrs = {}, returnElement = false) {
     if (typeof _icons[name] == "undefined") {
       console.error("JS icon missing: " + name);
       return name;
     }
+    const cacheKey = `${name}||${JSON.stringify(attrs)}`;
+    if (_iconsCache[cacheKey]) {
+      return returnElement
+        ? _iconsCache[cacheKey].cloneNode(true)
+        : _iconsCache[cacheKey].outerHTML;
+    }
+
+    const clone = _elementBase.cloneNode();
+    clone.insertAdjacentHTML('afterbegin', _icons[name]);
+    let element = clone.firstChild;
 
     // Add additional attributes
-    function addAttrs(_html, _attrs = {}) {
-      var mod = String(_html);
-      for (var attr in _attrs) {
-        if (Object.prototype.hasOwnProperty.call(_attrs, attr)) {
-          var sliceStart = mod.indexOf(" ");
-          var sliceEnd = sliceStart;
-          var value = _attrs[attr];
-          var regex = new RegExp(` ${attr}=(['"])([^\\1]+?)\\1`);
-          var existing = mod.match(regex);
-          if (existing) {
-            sliceStart = existing.index;
-            sliceEnd = sliceStart + existing[0].length;
-            value = existing[2] + " " + value;
-          }
-          mod = mod.slice(0, sliceStart) +
-              " " + attr + '="' + value + '"' +
-              mod.slice(sliceEnd);
+    function addAttrs(_element, _attrs = {}) {
+      Object.keys(_attrs).forEach(key => {
+        let newAttrs = _attrs[key].split(" ");
+        const oldAttrs = _element.getAttribute(key);
+        if (oldAttrs !== null) {
+          newAttrs = [...newAttrs, ...oldAttrs.split(" ")];
+          // Remove duplicate values
```

Comment:
You could use a Set to remove duplicate values as you compile them.

```JS
const newAttrsSet = new Set([...newAttrs, ...oldAttrs.split(" ")]);
//...
_element.setAttribute(key, Array.from(newAttrsSet).join(" "));
```

Reply:
Are there other attributes that require space separated strings other than class? If we're going to be working on an Element directly, I'd like Element to do as much work as possible. This actually makes me think we should be overriding the attributes instead of appending them (except for class). I know this is based on my code but I've since learned [how complicated this can get](https://github.com/johnnesky/imperative-html/blob/main/src/elements-base.ts#L3).

Reply:
Hmm, I would think that class could be the only one, only if there is some custom datasets but otherwise shouldn't be, ill try and adjust

Reply:
@crhallberg made some small changes where if the key is not class then just sets it instantly.